### PR TITLE
CI: add '[lint only]', '[docs only]' skip tags

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
https://github.com/scipy/scipy/pull/19491#issuecomment-1802656984 @mdhaber 

#### What does this implement/fix?
New '[lint only]' tag which will skip all CI workflows apart from `lint.yml`.

#### Additional information
This is my first time working on CI, sorry if I've missed anything obvious. This will obviously need to be documented as well.

### CODE REVIEW

**Comment by lucascolley** [comment]
Looks like it's worked! cc @j-bowhay also

File: .circleci/config.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ commands:
             export git_log=$(git log --max-count=1 --pretty=format:"%B" | tr "\n" " ")
             echo "Got commit message:"
             echo "${git_log}"
-            if [[ -v CIRCLE_PULL_REQUEST ]] && ([[ "$git_log" == *"[skip circle]"* ]] || [[ "$git_log" == *"[circle skip]"* ]]); then
+            if [[ -v CIRCLE_PULL_REQUEST ]] &&([[ "$git_log" == *"[skip circle]"* ]] || [[ "$git_log" == *"[circle skip]"* ]] || [[ "$git_log" == *"[lint only]"* ]]); then
```
Comment (lucascolley) [inline .circleci/config.yml:25]: whitespace deletion crept in here somehow

File: .github/workflows/circle_artifacts.yml [hunk 2]
```diff
@@ -2,7 +2,11 @@ on: [status]
 jobs:
   circleci_artifacts_redirector_job:
     runs-on: ubuntu-22.04
-    if: "github.repository == 'scipy/scipy' && !contains(github.event.head_commit.message, '[skip circle]')  && github.event.context == 'ci/circleci: build_docs'"
+    if: >-
+      "github.repository == 'scipy/scipy'
+      !contains(github.event.head_commit.message, '[skip circle]')
```
Comment (lucascolley) [inline .github/workflows/circle_artifacts.yml:7]: this is missing an `&&`

**Comment by lucascolley** [comment]
After fixing the syntax, it seems to be working for Circle, but actions are still running.

File: .github/workflows/circle_artifacts.yml [hunk 3]
```diff
@@ -2,7 +2,11 @@ on: [status]
 jobs:
   circleci_artifacts_redirector_job:
     runs-on: ubuntu-22.04
-    if: "github.repository == 'scipy/scipy' && !contains(github.event.head_commit.message, '[skip circle]')  && github.event.context == 'ci/circleci: build_docs'"
+    if: >
+      github.repository == 'scipy/scipy'
+      && !contains(github.event.head_commit.message, '[skip circle]')
+      && github.event.context == 'ci/circleci: build_docs'"
+      && !contains(github.event.head_commit.message, '[lint only]'))
```
Comment (lucascolley) [inline .github/workflows/circle_artifacts.yml:9]:
```suggestion
      && !contains(github.event.head_commit.message, '[lint only]')
```

**Comment by lucascolley** [comment]
Hmm I'm stuck, not sure what I've missed but the skips aren't working for the actions jobs.

File: .github/workflows/circle_artifacts.yml [hunk 4]
```diff
@@ -2,7 +2,11 @@ on: [status]
 jobs:
   circleci_artifacts_redirector_job:
     runs-on: ubuntu-22.04
-    if: "github.repository == 'scipy/scipy' && !contains(github.event.head_commit.message, '[skip circle]')  && github.event.context == 'ci/circleci: build_docs'"
+    if: >
+      github.repository == 'scipy/scipy'
+      && !contains(github.event.head_commit.message, '[skip circle]')
+      && github.event.context == 'ci/circleci: build_docs'"
```
Comment (lucascolley) [inline .github/workflows/circle_artifacts.yml:8]:
```suggestion
      && github.event.context == 'ci/circleci: build_docs'
```

**Comment by lucascolley** [comment]
Using the `get_commit_message` job from wheels it looks like I've got something working. Will come back and clean up

**Comment by lucascolley** [comment]
@mdhaber (after much struggle 😅) the [lint only] tag looks to be working now, see the first commit (although `circle_artifacts` looks like it's having a problem with `fatal: not a git repository (or any of the parent directories): .git`?).

@andyfaff does my diagnosis of the bug where `github.event.head_commit.message` doesn't work with `on: pull_request` look correct?

File: .github/workflows/circle_artifacts.yml [hunk 5]
```diff
@@ -2,7 +2,9 @@ on: [status]
 jobs:
   circleci_artifacts_redirector_job:
     runs-on: ubuntu-22.04
-    if: "github.repository == 'scipy/scipy' && !contains(github.event.head_commit.message, '[skip circle]')  && github.event.context == 'ci/circleci: build_docs'"
+    if: >
+      github.repository == 'scipy/scipy'
+      && github.event.context == 'ci/circleci: build_docs'
```
Comment (lucascolley) [inline .github/workflows/circle_artifacts.yml:7]: As far as I can tell, the behaviour is identical whether `&& !contains(github.event.head_commit.message, '[skip circle]')` is included or not.

**Comment by lucascolley** [comment]
It was suggested in the committee meeting that we might be able to take this 'get commit message' step out into an action which is used by each of the workflows. I'll try to figure that out when I find the time. Also, a [docs only] tag was suggested as a QOL improvement over [skip actions] [skip cirrus].

File: .cirrus.star [hunk 6]
```diff
@@ -39,7 +39,7 @@ def main(ctx):
     if PR < 0:
         return []
 
-    if "[skip cirrus]" in dct["message"] or "[skip ci]" in dct["message"]:
+    if "[skip cirrus]" in dct["message"] or "[skip ci]" in dct["message"] or "[lint only]" in dct["message"]:
```
Comment (lucascolley) [inline .cirrus.star:42]: I need to include [docs only] here

**Comment by lucascolley** [comment]
@mdhaber @melissawm this is ready for review now! We have a new workflow `commit_message.yml` which checks for [lint only] and [docs only] tags. The actions workflows each have a new job `Get commit message` which gets the output from `commit_message.yml` and skips accordingly. Circle and Cirrus configs are updated as needed too. The CI runs for each poke commit above exhibit the desired behaviour.

If this looks good, should we update the docs to point to '[docs only]' rather than '[skip actions] [skip cirrus]'? (the only difference there is that the linter runs, but that seems fine? we could change that but maybe not worth it)

File: .github/workflows/commit_message.yml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+          ref: ${{ github.event.pull_request.head.sha }}
+      - name: Check for skips
+        id: skip_check
+        run: |
+          set -xe
+          COMMIT_MSG=$(git log --no-merges -1)
+          RUN="1"
+          if [[ "$COMMIT_MSG" == *"[lint only]"* || "$COMMIT_MSG" == *"[docs only]"* ]]; then
+              RUN="0"
+          fi
```
Comment (lucascolley) [inline .github/workflows/commit_message.yml:31]: If we don't want the linter to run on '[docs only]', we could set `RUN="L"` for '[lint only]' then add a new job to the lint workflow to get the commit message and run if the output is `"1"` or `"L"`.

**Comment by mdhaber** [comment]
> If this looks good, should we update the docs to point to '[docs only]' rather than '[skip actions] [skip cirrus]'? 

Definitely. Thank you for pushing through to make this happen!

> The only difference there is that the linter runs...

Suppose we make `[docs only]` not run the linter.
Is it possible to use`[docs only]` and `[lint only]` to build docs and do linting?

**Comment by lucascolley** [comment]
> Is it possible to use [docs only] and [lint only] to build docs and do linting?

Yep, definitely possible. If we follow my inline suggestion above and make the logic such that it covers the combinations of the two tags correctly. The question is whether this extra complexity is worth it for the ~45 seconds saved on each [docs only] commit.

**Comment by mdhaber** [comment]
> The question is whether this extra complexity is worth it

Probably not. I'm happy either way.

File: .github/workflows/array_api.yml [hunk 8]
```diff
@@ -21,9 +21,17 @@ concurrency:
...
+    uses: ./.github/workflows/commit_message.yml
+
   pytorch_cpu:
     name: Linux PyTorch CPU
-    if: "github.repository == 'scipy/scipy' || github.repository == ''"
+    needs: get_commit_message
+    if: >
+      (github.repository == 'scipy/scipy'
+      && needs.get_commit_message.outputs.message == 1)
+      || github.repository == ''
```
[the remark this answers is not in the document]
Comment (melissawm) [reply .github/workflows/array_api.yml:34]: I'm all for consistency, but if this complicates things too much we can reconsider.

**Review by melissawm — COMMENTED** [review]
No strong opinions on the [docs only] including lint question. "Explicit is better than implicit" but "Simple is better than complex" 😄

**Comment by mdhaber** [comment]
Mind if I give this into Saturday for someone familiar with CI to take a look? If there are no takers and you think it's safe enough (i.e. these changes don't affect the way jobs run when these tags are absent), I'll be brave and merge that evening?

**Comment by lucascolley** [comment]
Sounds good. I'll get the documentation changes added before then.

**Comment by melissawm** [comment]
Please make sure to mention these options here: https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

**Comment by mdhaber** [comment]
Thanks @lucascolley! I'm going to try this out in gh-19510.

**Pull request merged**